libengine/dpengine/pixels.c

```diff
@@ -431,6 +433,137 @@ void DP_pixels15_to_8(DP_Pixel8 *dst, const DP_Pixel15 *src, int count)
+        p4 = _mm_slli_si128(p4, 3);
+
+        __m128i combined =
+            _mm_or_si128(p1, _mm_or_si128(p2, _mm_or_si128(p3, p4)));
+
+        __m128i out = _mm_shuffle_epi8(
+            combined, _mm_setr_epi8(0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14, 3,
+                                    7, 11, 15));
+
+        _mm_store_si128((__m128i *)&dst[i], out);
+    }
+}
+DP_TARGET_END
+
+DP_TARGET_BEGIN("avx2")
+static void DP_pixels15_to_8_avx2(DP_Pixel8 *dst, const DP_Pixel15 *src,
+                                  int count)
+{
+    assert(count % 8 == 0);
+
+    for (int i = 0; i < count; i += 8) {
+        __m256i source1 = _mm256_loadu_si256((__m256i *)&src[i]);
+        __m256i source2 = _mm256_loadu_si256((__m256i *)&src[i + 4]);
+
+        __m256i p1 = _mm256_and_si256(source1, _mm256_set1_epi32(0xFFFF));
+        __m256i p2 = _mm256_srli_epi32(
+            _mm256_and_si256(source1, _mm256_set1_epi32(0xFFFF0000)), 16);
+        __m256i p3 = _mm256_and_si256(source2, _mm256_set1_epi32(0xFFFF));
+        __m256i p4 = _mm256_srli_epi32(
+            _mm256_and_si256(source2, _mm256_set1_epi32(0xFFFF0000)), 16);
+
+        p1 = _mm256_srli_epi32(_mm256_mullo_epi32(p1, _mm256_set1_epi32(255)),
+                               15);
+        p2 = _mm256_srli_epi32(_mm256_mullo_epi32(p2, _mm256_set1_epi32(255)),
+                               15);
+        p3 = _mm256_srli_epi32(_mm256_mullo_epi32(p3, _mm256_set1_epi32(255)),
+                               15);
+        p4 = _mm256_srli_epi32(_mm256_mullo_epi32(p4, _mm256_set1_epi32(255)),
+                               15);
+
+        p1 = _mm256_shuffle_epi8(
+            p1, _mm256_setr_epi8(0, 0x80, 4, 0x80, 0x80, 0x80, 0x80, 0x80, 8,
+                                 0x80, 12, 0x80, 0x80, 0x80, 0x80, 0x80, 0,
+                                 0x80, 4, 0x80, 0x80, 0x80, 0x80, 0x80, 8, 0x80,
+                                 12, 0x80, 0x80, 0x80, 0x80, 0x80));
+        p2 = _mm256_shuffle_epi8(
+            p2, _mm256_setr_epi8(0x80, 0, 0x80, 4, 0x80, 0x80, 0x80, 0x80, 0x80,
+                                 8, 0x80, 12, 0x80, 0x80, 0x80, 0x80, 0x80, 0,
+                                 0x80, 4, 0x80, 0x80, 0x80, 0x80, 0x80, 8, 0x80,
+                                 12, 0x80, 0x80, 0x80, 0x80));
+        p3 = _mm256_shuffle_epi8(
+            p3, _mm256_setr_epi8(0x80, 0x80, 0x80, 0x80, 0, 0x80, 4, 0x80, 0x80,
+                                 0x80, 0x80, 0x80, 8, 0x80, 12, 0x80, 0x80,
+                                 0x80, 0x80, 0x80, 0, 0x80, 4, 0x80, 0x80, 0x80,
+                                 0x80, 0x80, 8, 0x80, 12, 0x80));
+        p4 = _mm256_shuffle_epi8(
+            p4, _mm256_setr_epi8(0x80, 0x80, 0x80, 0x80, 0x80, 0, 0x80, 4, 0x80,
+                                 0x80, 0x80, 0x80, 0x80, 8, 0x80, 12, 0x80,
+                                 0x80, 0x80, 0x80, 0x80, 0, 0x80, 4, 0x80, 0x80,
+                                 0x80, 0x80, 0x80, 8, 0x80, 12));
```

Comment:
Eighth bit needs a cast to avoid warnings about unsigned to char conversion: `(char)0x80`

Putting it into a `char bit8 = (char)0x80` or a `#define BIT8 ((char)0x80)` or something might be less annoying.